File: platform/Emscripten/src/benchmark_timer.js
const { performance } = require("node:perf_hooks");

const WARMUP_ROUNDS = 10;

function measure(iterations, fn) {
  for (let index = 0; index < WARMUP_ROUNDS; index += 1) {
    fn();
  }

  let totalMs = 0;
  let minMs = Number.POSITIVE_INFINITY;
  let maxMs = 0;

  for (let index = 0; index < iterations; index += 1) {
    const begin = performance.now();
    fn();
    const elapsedMs = performance.now() - begin;
    totalMs += elapsedMs;
    minMs = Math.min(minMs, elapsedMs);
    maxMs = Math.max(maxMs, elapsedMs);
  }

  return {
    averageMs: totalMs / iterations,
    minMs,
    maxMs,
  };
}

module.exports = {
  measure,
};

File: platform/Android/app/src/main/java/com/ffibb/android/BenchmarkTimer.java
package com.ffibb.android;

final class BenchmarkTimer {
  interface Task {
    long run();
  }

  private BenchmarkTimer() {
  }

  static BenchmarkMeasurement measure(int iterations, int warmupRounds, Task task) {
    for (int warmup = 0; warmup < warmupRounds; warmup++) {
      BenchmarkSink.add(task.run());
    }

    double totalMs = 0.0;
    double minMs = Double.MAX_VALUE;
    double maxMs = 0.0;

    for (int iteration = 0; iteration < iterations; iteration++) {
      long start = System.nanoTime();
      long sinkValue = task.run();
      long end = System.nanoTime();
      BenchmarkSink.add(sinkValue);

      double elapsedMs = (end - start) / 1_000_000.0;
      totalMs += elapsedMs;
      minMs = Math.min(minMs, elapsedMs);
      maxMs = Math.max(maxMs, elapsedMs);
    }

    return new BenchmarkMeasurement(totalMs / iterations, minMs, maxMs);
  }
}
